cmd/rekor-cli/app/timestamp.go

```diff
@@ -0,0 +1,251 @@
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+type fileFlag struct {
+	value string
+}
+
+func (f *fileFlag) String() string {
+	return f.value
+}
+
+func (f *fileFlag) Set(s string) error {
+	if s == "" {
+		return errors.New("flag must be specified")
+	}
+	if _, err := os.Stat(filepath.Clean(s)); os.IsNotExist(err) {
+		return err
+	}
+	f.value = s
+	return nil
+}
+
+func (f *fileFlag) Type() string {
+	return "fileFlag"
+}
+
+type oidFlag struct {
+	value asn1.ObjectIdentifier
+}
+
+func (f *oidFlag) String() string {
+	return f.value.String()
+}
+
+func (f *oidFlag) Set(s string) error {
+	parts := strings.Split(s, ".")
+	f.value = make(asn1.ObjectIdentifier, len(parts))
+	for i, part := range parts {
+		num, err := strconv.Atoi(part)
+		if err != nil {
+			return errors.New("error parsing OID")
+		}
+		f.value[i] = num
+	}
+	return nil
+}
+
+func (f *oidFlag) Type() string {
+	return "oidFlag"
+}
+
+func addTimestampFlags(cmd *cobra.Command) error {
+	cmd.Flags().Var(&fileFlag{}, "artifact", "path to an artifact to timestamp")
+	cmd.Flags().Var(&uuidFlag{}, "artifact-hash", "hex encoded SHA256 hash of the the artifact to timestamp")
+	cmd.Flags().Bool("nonce", false, "specify a pseudo-random nonce in the request")
+	cmd.Flags().Var(&oidFlag{}, "tsa-policy", "optional dotted OID notation for the policy that the TSA should use to create the response")
+
+	cmd.Flags().String("out", "response.tsr", "path to a file to write response.")
+
+	// TODO: Add a flag to upload a pre-formed timestamp response to the log.
+	// TODO: Add a flag to indicate a JSON formatted timestamp request/response.
+	return nil
+}
+
+func validateTimestampFlags() error {
+	artifactStr := viper.GetString("artifact")
+	digestStr := viper.GetString("artifact-hash")
+
+	if artifactStr == "" && digestStr == "" {
+		return errors.New("artifact or hash to timestamp must be specified")
+	}
+
+	if digestStr != "" {
+		digest := uuidFlag{}
+		if err := digest.Set(digestStr); err != nil {
+			return err
+		}
+	}
+
+	policyStr := viper.GetString("tsa-policy")
+	if policyStr != "" {
+		oid := oidFlag{}
+		if err := oid.Set(policyStr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createRequestFromFlags() (*pkcs9.TimeStampReq, error) {
+	var timestampReq *pkcs9.TimeStampReq
+	digestStr := viper.GetString("artifact-hash")
+	policyStr := viper.GetString("tsa-policy")
+
+	opts := util.TimestampRequestOptions{
+		// Always use a SHA256 right now.
+		Hash: crypto.SHA256,
+	}
+	if policyStr != "" {
+		oid := oidFlag{}
+		if err := oid.Set(policyStr); err != nil {
+			return nil, err
+		}
+		opts.TSAPolicyOid = oid.value
+	}
+	if viper.GetBool("nonce") {
+		opts.Nonce = x509tools.MakeSerial()
+	}
+
+	var digest []byte
+	if digestStr != "" {
+		decoded, err := hex.DecodeString(digestStr)
+		if err != nil {
+			return nil, err
+		}
+		digest = decoded
+	}
+	if digestStr == "" {
+		artifactStr := viper.GetString("artifact")
```

Comment:
Yeah, I thought about that, can do in a followup -- here's my question about it: I want to server to accept a TSQ, so this would mean the client would need to fetch the URL which I wanted to avoid. Instead, I allowed the artifact-hash which is basically the same.

Otherwise, I could add this as an option for the JSON formatted and expose a URL field that the server can fetch.